Optimize away `z_bytes_copy_from_buf` in `rmw_publish`

fuzzypixelz opened the pull request.

fuzzypixelz committed:
Optimize `PublisherData::publish`

1. `z_bytes_copy_from_buf` is replaced by `z_bytes_from_buf`.
2. A pool of serialization buffers is introduced to reduce
   allocations and improve cache locality.

fuzzypixelz committed:
Use temporary zenoh-c patch

YuanYuYuan commented:
Hi @fuzzypixelz ,

1. It seems that the memory allocation is infallible. We may return the error if there is not sufficient memory.
2. It's better to attach the zenoh_rmw_eval results to ensure this PR also passes most of the tests.

Commit:
Wip

DONE:
- Fix std::string constructor call in rmw_subscription_data.cpp
- Move serialization buffer pool to the context

TODO:
- Remove printf calls
- Remove sanitizer usage
- Fix formatting
- Resolve FIXMEs in BufferPool impl

Commit:
Wip

Commit:
Wip
